src/equations/shallow_water_multilayer_1d.jl

```diff
@@ -79,19 +89,32 @@ end
 # The reference total water height H0 defaults to 0.0 but is used for the "lake-at-rest"
 # well-balancedness test cases. 
 function ShallowWaterMultiLayerEquations1D(; gravity_constant,
-                                           H0 = zero(gravity_constant), rhos)
+                                           H0 = zero(gravity_constant),
+                                           threshold_limiter = nothing,
+                                           threshold_desingularization = nothing, rhos)
 
     # Promote all variables to a common type
     _rhos = promote(rhos...)
     RealT = promote_type(eltype(_rhos), eltype(gravity_constant), eltype(H0))
     __rhos = SVector(map(RealT, _rhos))
 
+    # Set default values for thresholds
+    if threshold_limiter === nothing
+        threshold_limiter = 5 * eps(RealT)
+    end
+    if threshold_desingularization === nothing
+        threshold_desingularization = 1e-10
```

Comment:
This would not work for single precision computations. What would an analogous value for this threshold be in single precision? Then we could set it using a similar tactic as `<some constant> * eps(RealT)`.

Reply:
The choice for this threshold is only based on heuristics. 
It should be chosen small enough as to not influence the solution, but also big enough to not introduce numerical problems. I think we would have to experiment a bit with single precision, to determine a good value for this.
Rewriting this in terms of a fixed constant gives `5e5*eps(Float64)=1.11e-10` and `5e5*eps(Float32)=0.0596...`, which seems very high to me in the single precision case.
Maybe it's better to either choose fixed values depending on number format or just enforce users to provide a threshold until we decide on a good threshold:
```julia
    if threshold_desingularization === nothing
        if RealT <: Float64
            threshold_desingularization = 1e-10
        else
            throw(ArgumentError("threshold_desingularization must be set manually for non-Float64 types"))
        end
    end
```

Reply:
> It should be chosen small enough as to not influence the solution, but also big enough to not introduce numerical problems.

This is classic for positivity preserving methods :)

I am okay with your current suggestion because the development of TrixiShallowWater will focus on CPUs. GPU support is something to push to the future. From the old code of Niklas he set the dry tolerance to be `1e-4` for the GPU-OCCA implementation. @ranocha Do you have any experience with SWE + wet/dry on GPUs (or suggestions on the best way to handle this)?

Reply:
Are there other locations in the code where you hard-code floating point constants? With an `eps`? If yes, then IMHO you don't need to find a solution for this issue right here right now, but I'd instead record an issue and defer the decision until it becomes relevant.

Other than that, instead of an `if` condition, I'd use a function:
```julia
default_threshold_desingularization(::Type{Float64)) = 1e-10
# default_threshold_desingularization(::Type{Float32)) = 1e-8
```
That way you can be sure that there will be an error if someone tries to use a different type for which no threshold has been found yet, but at the same time making it easy to provide one (e.g., if you decide you want to use AD or Measurements.jl and suddenly have a different type you need to support) without having to change the Trixi.jl source code.

Reply:
There is another hard-coded threshold that is used for the [`IndicatorHennemannGassnerShallowWater`](https://github.com/trixi-framework/TrixiShallowWater.jl/blob/78e4f35420b58e85016a90fefcf8b65c7315d4ee/src/solvers/indicators_1d.jl#L48) that is set to `threshold_partially_wet = 1e-4`, which might be okay for single precision. This value is currently not adjustable, as it is not a field of the equation struct. I think we should also make this part of the equation struct, to provide an overview of all manually set parameters at the same place and make it adjustable if needed.

Since the computations can be very sensitive to the thresholds and we don't use single precision right now, I would also prefer to create an issue for now and only provide default values once we have done some testing in single precision. In the meantime we still provide the option to set these parameters explicitly if needed.

@sloede is there any downside with using the `if` condition? I think it would be good to provide some guidance with the error message.

Reply:
The downside is that someone who quickly wants to use a different data type will now have to change TrixiSW itself to make it work. With the functions, you can just define the function externally and do not need to change the source code of TrixiSW itself.

If you want a nicer error message, you can add a function
```julia
default_threshold_desingularization(catchall) = throw(ArgumentError("..."))
```
that gets called if no appropriate function for a data type is found.

Reply:
The thresholds can currently be set when initializing the equations, without changing TrixiSW, as they are part of the equation struct. But I guess doing with functions is still nicer and offers additional flexibility.

Reply:
I have opened issue #41 to document and discuss how to change the thresholds to support single precision.

Reply:
> > It should be chosen small enough as to not influence the solution, but also big enough to not introduce numerical problems.
> 
> This is classic for positivity preserving methods :)
> 
> I am okay with your current suggestion because the development of TrixiShallowWater will focus on CPUs. GPU support is something to push to the future. From the old code of Niklas he set the dry tolerance to be `1e-4` for the GPU-OCCA implementation. @ranocha Do you have any experience with SWE + wet/dry on GPUs (or suggestions on the best way to handle this)?

Sorry, I missed this message. The approach you have discussed above sounds good to me - default handling via a function specializing on the type, gathering more experience first.